packages/react_rearch/lib/src/widgets/consumer.dart

```diff
@@ -0,0 +1,158 @@
+part of '../widgets.dart';
+
+abstract class RearchComponent extends Component2 {
+  bool _building = false;
+
+  String get debugName;
+
+  bool get debug => false;
+
+  void _debug(String msg) {
+    if (!debug) return;
+    // ignore: avoid_print
+    print('<$debugName> -- $msg');
+  }
+
+  final willUnmountListeners = <SideEffectApiCallback>{};
+
+  final sideEffectData = <Object?>[];
+
+  /// Represents a [Set] of functions that remove a dependency on a [Capsule].
+  final dependencyDisposers = <void Function()>{};
+
+  /// Clears out the [Capsule] dependencies of this [RearchComponent].
+  void clearDependencies() {
+    for (final dispose in dependencyDisposers) {
+      dispose();
+    }
+    dependencyDisposers.clear();
+  }
+
+  @override
+  void componentDidMount() {
+    _debug('componentDidMount()');
+    super.componentDidMount();
+  }
+
+  @override
+  void componentDidUpdate(
+    Map<dynamic, dynamic> prevProps,
+    Map<dynamic, dynamic> prevState, [
+    dynamic snapshot,
+  ]) {
+    _debug('componentDidUpdate()');
+    super.componentDidUpdate(prevProps, prevState, snapshot);
+  }
+
+  @override
+  void componentWillUnmount() {
+    _debug('componentWillUnmount()');
+
+    for (final listener in willUnmountListeners) {
+      listener();
+    }
+
+    clearDependencies();
+
+    // Clean up after any side effects to avoid possible leaks
+    willUnmountListeners.clear();
+
+    super.componentWillUnmount();
+  }
+
+  @override
+  ReactNode render() {
+    _debug('render()');
+
+    _building = true;
+
+    // Clears the old dependencies (which will be repopulated via WidgetHandle)
+    clearDependencies();
+
+    final res = build(
+      _ComponentHandleImpl(
+        _ComponentSideEffectApiProxyImpl(this),
+        topLevelCapsuleContainer,
+      ),
+    );
+
+    _building = false;
+
+    return res;
+  }
+
+  ReactNode build(ComponentHandle use);
+}
+
+/// This is needed so that [ComponentSideEffectApi.rebuild] doesn't conflict
+/// with [Component2.forceUpdate].
+class _ComponentSideEffectApiProxyImpl implements ComponentSideEffectApi {
```

Comment:
This was needed in Flutter ReArch since elements have a `rebuild` method, which conflicted with `api.rebuild`.

There isn't a rebuild method in components as far as I can tell, so instead the `RearchComponent` should likely implement `ComponentSideEffectApi` instead.

Reply:
Done